lib/phoenix_live_view/async.ex

```diff
@@ -0,0 +1,215 @@
+      when is_atom(key) and is_function(func, 0) do
+    run_async_task(socket, key, func, :start)
+  end
+
+  def assign_async(%Socket{} = socket, key_or_keys, func)
+      when (is_atom(key_or_keys) or is_list(key_or_keys)) and
+             is_function(func, 0) do
+    keys = List.wrap(key_or_keys)
+
+    # verifies result inside task
+    wrapped_func = fn ->
+      case func.() do
+        {:ok, %{} = assigns} ->
+          if Enum.find(keys, &(not is_map_key(assigns, &1))) do
+            raise ArgumentError, """
+            expected assign_async to return map of assigns for all keys
+            in #{inspect(keys)}, but got: #{inspect(assigns)}
+            """
+          else
+            {:ok, assigns}
+          end
+
+        {:error, reason} ->
+          {:error, reason}
+
+        other ->
+          raise ArgumentError, """
+          expected assign_async to return {:ok, map} of
+          assigns for #{inspect(keys)} or {:error, reason}, got: #{inspect(other)}
+          """
+      end
+    end
+
+    new_assigns =
+      Enum.map(keys, fn key ->
+        case socket.assigns do
+          %{^key => %AsyncResult{ok?: true} = existing} ->
+            {key, AsyncResult.loading(existing, keys)}
+
+          %{} ->
+            {key, AsyncResult.loading(keys)}
+        end
+      end)
+
+    socket
+    |> Phoenix.Component.assign(new_assigns)
+    |> run_async_task(keys, wrapped_func, :assign)
+  end
+
+  defp run_async_task(%Socket{} = socket, key, func, kind) do
+    if Phoenix.LiveView.connected?(socket) do
+      lv_pid = self()
+      cid = cid(socket)
+      {:ok, pid} = Task.start_link(fn -> do_async(lv_pid, cid, key, func, kind) end)
+
+      ref =
+        :erlang.monitor(:process, pid, alias: :reply_demonitor, tag: {__MODULE__, key, cid, kind})
+
+      send(pid, {:context, ref})
+
+      update_private_async(socket, &Map.put(&1, key, {ref, pid, kind}))
+    else
+      socket
+    end
+  end
+
+  defp do_async(lv_pid, cid, key, func, async_kind) do
+    receive do
+      {:context, ref} ->
+        try do
+          result = func.()
+          Channel.report_async_result(ref, async_kind, ref, cid, key, {:ok, result})
+        catch
+          catch_kind, reason ->
+            Process.unlink(lv_pid)
+            caught_result = to_exit(catch_kind, reason, __STACKTRACE__)
+            Channel.report_async_result(ref, async_kind, ref, cid, key, caught_result)
+            :erlang.raise(catch_kind, reason, __STACKTRACE__)
+        end
+    end
+  end
+
+  def cancel_async(%Socket{} = socket, %AsyncResult{} = result, reason) do
+    case result do
+      %AsyncResult{loading: keys} when is_list(keys) ->
+        new_assigns = for key <- keys, do: {key, AsyncResult.failed(result, {:exit, reason})}
+
+        socket
+        |> Phoenix.Component.assign(new_assigns)
+        |> cancel_async(keys, reason)
+
+      %AsyncResult{} ->
+        socket
+    end
+  end
+
+  def cancel_async(%Socket{} = socket, key, _reason) do
```

Comment:
Shouldn't we support a list of keys? I don't see where that is handled. :( But also note that if I do:

    cancel_async(socket, :key, ...)

We won't set the async_result under socket.assign.key to failed.

Reply:
the "key" is whatever the user passes to `start_async`, so `:foo` or `[:foo, :bar, :baz]`. So the list of keys is supported because the key is a list of keys in some cases. I will bump the docs around this.